src/MixedMatrix.cpp

```diff
@@ -15,221 +15,311 @@
-                         std::unique_ptr<mfem::SparseMatrix> W,
-                         const mfem::HypreParMatrix& edge_d_td)
-    :
-    M_(std::move(M)),
-    D_(std::move(D)),
-    W_(std::move(W)),
-    edge_d_td_(&edge_d_td),
-    edge_td_d_(edge_d_td_->Transpose())
+
+int MixedMatrix::Rows() const
 {
-    GenerateRowStarts();
+    return D_local_.Rows() + D_local_.Cols();
 }
 
-/// @todo better documentation of the 1/-1 issue, make it optional?
-void MixedMatrix::Init(const mfem::SparseMatrix& vertex_edge,
-                       const mfem::Vector& weight,
-                       const mfem::SparseMatrix& w_block)
+int MixedMatrix::Cols() const
 {
+    return D_local_.Rows() + D_local_.Cols();
+}
 
+int MixedMatrix::GlobalRows() const
+{
+    return D_global_.GlobalRows() + D_global_.GlobalCols();
+}
 
-    const mfem::HypreParMatrix& edge_d_td(*edge_d_td_);
-    const int nvertices = vertex_edge.Height();
-    const int nedges = vertex_edge.Width();
+int MixedMatrix::GlobalCols() const
+{
+    return D_global_.GlobalRows() + D_global_.GlobalCols();
+}
 
-    int* M_fine_i = new int [nedges + 1];
-    int* M_fine_j = new int [nedges];
-    double* M_fine_data = new double [nedges];
-    std::iota(M_fine_i, M_fine_i + nedges + 1, 0);
-    std::iota(M_fine_j, M_fine_j + nedges, 0);
-    std::copy_n(weight.GetData(), nedges, M_fine_data);
+int MixedMatrix::NNZ() const
+{
+    return M_local_.nnz() + (2 * D_local_.nnz())
+           + W_local_.nnz();
+}
 
-    for (int i = 0; i < nedges; i++)
-    {
-        assert(M_fine_data[i] != 0.0);
-        M_fine_data[i] = 1.0 / std::fabs(M_fine_data[i]);
-    }
+int MixedMatrix::GlobalNNZ() const
+{
+    return M_global_.nnz() + (2 * D_global_.nnz())
+           + W_global_.nnz();
+}
 
-    M_ = make_unique<mfem::SparseMatrix>(M_fine_i, M_fine_j, M_fine_data,
-                                         nedges, nedges);
+bool MixedMatrix::CheckW() const
+{
+    int local_size = W_global_.Rows();
+    int global_size;
+
+    MPI_Allreduce(&local_size, &global_size, 1, MPI_INT, MPI_MAX, D_global_.GetComm());
+
+    const double zero_tol = 1e-6;
+
+    return global_size > 0 && W_global_.MaxNorm() > zero_tol;
+}
+
+
+ParMatrix MixedMatrix::ToPrimal() const
+{
+    assert(M_global_.Cols() == D_global_.Cols());
+    assert(M_global_.Rows() == D_global_.Cols());
+
+    ParMatrix MinvDT = D_global_.Transpose();
+    MinvDT.InverseScaleRows(M_global_.GetDiag().GetDiag());
 
-    if (w_block.Height() == nvertices && w_block.Width() == nvertices)
+    ParMatrix A = D_global_.Mult(MinvDT);
+
+    if (CheckW())
     {
-        W_ = make_unique<mfem::SparseMatrix>(w_block);
-        (*W_) *= -1.0;
+        A = ParAdd(A, W_global_);
     }
 
-    // Nonzero row of edge_owned means the edge is owned by the local proc
-    mfem::SparseMatrix edge_owned;
-    edge_d_td.GetDiag(edge_owned);
+    return A;
+}
 
-    mfem::SparseMatrix graphDT(smoothg::Transpose(vertex_edge));
+void MixedMatrix::AssembleM()
```

Comment:
This can simply call `AssembleM(const std::vector<double>& agg_weight)`? I understand this way will need to create an unnecessary const vector of 1s, but if we are willing to store a diagonal element matrix as dense matrix, this is extra vector of 1s won't really affect the performance? Especially I believe we dont need this function in loops (Monto Carlo or nonlinear solve).

Reply:
Yeah the unweighted assemble is only called once during the initial setup phase.